File: FallDetectionAppJava/app/src/main/java/com/example/falldetectionappjava/MainActivity.java
package com.example.falldetectionappjava;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;
import android.telephony.SmsManager;
import android.content.pm.PackageManager;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import android.Manifest;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

import android.content.Intent;
public class MainActivity extends AppCompatActivity implements SensorEventListener {

    private static final int SMS_PERMISSION_CODE = 100;
    private EditText phoneNumberInput;
    private Button testButton;

    private SensorManager sensorManager;
    private Sensor accelerometer;
    private static final float FALL_THRESHOLD = 25.0f; // ajusta según pruebas
    private boolean fallDetected = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        phoneNumberInput = findViewById(R.id.phoneNumberInput);
        testButton = findViewById(R.id.testButton);

        testButton.setOnClickListener(v -> {
            if (checkSmsPermission()) {
                sendSms("Prueba manual: ¡Caída detectada!");
            } else {
                requestSmsPermission();
            }
        });

        // Inicializamos SensorManager
        sensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);
        if (sensorManager != null) {
            accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (accelerometer != null) {
            sensorManager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_NORMAL);
        }
    }
    @Override
    protected void onPause() {
        super.onPause();
        sensorManager.unregisterListener(this);
    }
    @Override
    public void onSensorChanged(SensorEvent event) {
        if (event.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
            float x = event.values[0];
            float y = event.values[1];
            float z = event.values[2];

            // Magnitud de la aceleración
            double magnitude = Math.sqrt(x * x + y * y + z * z);

            if (magnitude > FALL_THRESHOLD && !fallDetected) {
                fallDetected = true;

                // Lanzar la pantalla de confirmación
                String phone = phoneNumberInput.getText().toString().trim();
                Intent intent = new Intent(MainActivity.this, EmergencyActivity.class);
                intent.putExtra("PHONE_NUMBER", phone);
                startActivity(intent);

                // Reiniciar detección tras un tiempo
                phoneNumberInput.postDelayed(() -> fallDetected = false, 5000);
            }

        }
    }
    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // No usamos
    }


    private void sendSms(String message) {
        String phoneNumber = phoneNumberInput.getText().toString().trim();
        if (phoneNumber.isEmpty()) {
            Toast.makeText(this, "Ingresa un número de emergencia", Toast.LENGTH_SHORT).show();
            return;
        }
        try {
            SmsManager smsManager = SmsManager.getDefault();
            smsManager.sendTextMessage(phoneNumber, null, message, null, null);
            Toast.makeText(this, "SMS enviado a " + phoneNumber, Toast.LENGTH_SHORT).show();
        } catch (Exception e) {
            Toast.makeText(this, "Error enviando SMS: " + e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    private boolean checkSmsPermission() {
        return ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS)
                == PackageManager.PERMISSION_GRANTED;
    }

    private void requestSmsPermission() {
        ActivityCompat.requestPermissions(this,
                new String[]{Manifest.permission.SEND_SMS},
                SMS_PERMISSION_CODE);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions,
                                           int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == SMS_PERMISSION_CODE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                sendSms("Permiso concedido, reenviando alerta de caída");
            } else {
                Toast.makeText(this, "Permiso de SMS denegado", Toast.LENGTH_SHORT).show();
            }
        }
    }
}

File: FallDetectionApp/app/src/main/java/com/example/falldetectionapp/MainActivity.kt
package com.example.falldetectionapp  // <-- nombre de paquete

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.telephony.SmsManager
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat
import kotlin.math.sqrt
import com.example.falldetectionapp.ui.theme.FallDetectionAppTheme  // <-- theme adaptado

class MainActivity : ComponentActivity(), SensorEventListener {
    private lateinit var sensorManager: SensorManager
    private var accelerometer: Sensor? = null

    private var _magnitude = mutableStateOf(0f)
    private var _fallDetected = mutableStateOf(false)

    // Variables para lógica de caída
    private var fallStage = 0 // 0=nada, 1=caída libre detectada

    // Temporizador para SMS
    private var handler: Handler? = null

    // Registrar permisos de SMS
    private val requestSmsPermissionLauncher =
        registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
            if (!granted) {
                Toast.makeText(this, "Permiso SMS denegado, no se podrá enviar alerta", Toast.LENGTH_SHORT).show()
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Preparar sensores
        sensorManager = getSystemService(SENSOR_SERVICE) as SensorManager
        accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)

        // Preparar temporizador
        handler = Handler(Looper.getMainLooper())

        // Pedir permiso SMS
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS)
            != PackageManager.PERMISSION_GRANTED
        ) {
            requestSmsPermissionLauncher.launch(Manifest.permission.SEND_SMS)
        }

        setContent {
            var emergencyNumber by remember { mutableStateOf(loadEmergencyNumber() ?: "") }

            FallDetectionUI(
                magnitude = _magnitude.value,
                fallDetected = _fallDetected.value,
                emergencyNumber = emergencyNumber,
                onEmergencyNumberChange = { emergencyNumber = it },
                onSaveNumber = {
                    saveEmergencyNumber(emergencyNumber)
                    Toast.makeText(this, "Número de emergencia guardado", Toast.LENGTH_SHORT).show()
                },
                onCancel = {
                    _fallDetected.value = false
                    Toast.makeText(this, "Alerta cancelada", Toast.LENGTH_SHORT).show()
                }
            )
        }
    }
    override fun onResume() {
        super.onResume()
        accelerometer?.also { accel ->
            sensorManager.registerListener(this, accel, SensorManager.SENSOR_DELAY_NORMAL)
        }
    }

    override fun onPause() {
        super.onPause()
        sensorManager.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent?) {
        event?.let {
            if (it.sensor.type == Sensor.TYPE_ACCELEROMETER) {
                val x = it.values[0]
                val y = it.values[1]
                val z = it.values[2]
                val magnitude = sqrt(x * x + y * y + z * z) //formula usada para detectar velocidad
                _magnitude.value = magnitude

                //Configuración de valores para detección de caída

                when (fallStage) {
                    0 -> if (magnitude < 2) fallStage = 1 // caída libre detectada
                    1 -> if (magnitude > 15) {
                        fallStage = 2
                        _fallDetected.value = true
                        startFallTimer()
                    }
                }
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // No necesario en este ejemplo
    }

    private fun startFallTimer() {
        handler?.postDelayed({
            if (_fallDetected.value) {
                val emergencyNumber = loadEmergencyNumber()
                if (!emergencyNumber.isNullOrBlank()) {
                    sendEmergencySms(emergencyNumber)
                } else {
                    Toast.makeText(this, "No hay número de emergencia configurado", Toast.LENGTH_SHORT).show()
                }
                _fallDetected.value = false
            }
        }, 30000)
    }

    private fun sendEmergencySms(phoneNumber: String) {
        try {
            val smsManager = SmsManager.getDefault()
            smsManager.sendTextMessage(
                phoneNumber,
                null,
                "Posible caída detectada. Necesito ayuda.",
                null,
                null
            )
            Toast.makeText(this, "SMS de emergencia enviado a $phoneNumber", Toast.LENGTH_SHORT).show()
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(this, "Error al enviar SMS", Toast.LENGTH_SHORT).show()
        }
    }
    // --- Persistencia con SharedPreferences ---
    private fun saveEmergencyNumber(number: String) {
        val prefs = getSharedPreferences("fall_prefs", Context.MODE_PRIVATE)
        prefs.edit().putString("emergency_number", number).apply()
    }

    private fun loadEmergencyNumber(): String? {
        val prefs = getSharedPreferences("fall_prefs", Context.MODE_PRIVATE)
        return prefs.getString("emergency_number", null)
    }
}

@Composable
fun FallDetectionUI(
    magnitude: Float,
    fallDetected: Boolean,
    emergencyNumber: String,
    onEmergencyNumberChange: (String) -> Unit,
    onSaveNumber: () -> Unit,
    onCancel: () -> Unit
) {
    Surface(
        modifier = Modifier.fillMaxSize(),
        color = MaterialTheme.colorScheme.background
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(16.dp),
            verticalArrangement = Arrangement.Top,
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text("Magnitud actual: ${"%.2f".format(magnitude)} m/s²")
            Spacer(modifier = Modifier.height(20.dp))

            OutlinedTextField(
                value = emergencyNumber,
                onValueChange = onEmergencyNumberChange,
                label = { Text("Número de emergencia") }
            )
            Spacer(modifier = Modifier.height(8.dp))
            Button(onClick = onSaveNumber) {
                Text("Guardar número")
            }

            Spacer(modifier = Modifier.height(30.dp))

            if (fallDetected) {
                Text("¡Posible caída detectada!")
                Spacer(modifier = Modifier.height(10.dp))
                Button(onClick = onCancel) {
                    Text("Estoy bien, cancelar alerta")
                }
            } else {
                Text("Sin caídas detectadas")
            }
        }
    }
}